### ISSUE TITLE
Second version of the pipeline

### ISSUE THREAD

**Issue by manulera** [manulera/genestorian_data_refinement#22]
In this new version, we should aim to produce two json files:

`strains.json` contains a list of objects representing strains (one for each row in `strains.tsv`).

For a line in strains.csv like this:

```tsv
Sample Name	Genotype
ID123	h- cls1-36 ase1D:NatMx
```

```javascript
[
    {
        id: 'ID123', // This should be a unique identifier for the strain in the file, the `sample name` field in the dey lab.
        genoytpe: 'h- cls1-36 ase1D:NatMx', // The genotype as listed in the tsv file
        mating_type: 'h-', // This can only be h- / h+ / h90
        alleles: [
            'cls1-36',
            'ase1D:NatMx'
        ]
    }
]

```

An `alleles.json` file that contains a list of found alleles, and what you have substituted in them. There should be only one entry per allele name: if the same allele name is present twice (e.g. if two strains have `ase1D:NatMX` in them), there should be only a single entry for the allele in `alleles.json`.

```javascript
[
    {
        name: 'cls1-36', // The name of the allele as written in the genotype
        pattern: 'ALLELE', 
        allele_features: [
            {
                name: 'cls1-36',
                feature_type: 'ALLELE'
            }
        ]
    },
    {
        name: `ase1D:NatMx`,
        pattern: 'GENEd-MARKER',
        allele_features: [
            {
                name: 'ase1',
                feature_type: 'GENE'
            },
            {
                name: 'NatMx',
                feature_type: 'MARKER'
            }
        ]
    }
]
```

How you should go about this:

* [x] To make sure that your function works for the example I provided, create an example `tsv` file with this content and test your code there:
    ```tsv
    Sample Name	Genotype
    ID123	h- cls1-36 ase1D:NatMx
    ```
* [x] First, create the function that produces `strains.json`, since this does not require finding any pattern.
* [x] Then, create the first version of the function that produces `alleles.json`, but only with the `name` field.
  * Since there should be only one entry per allele, to avoid repetition, you can store your alleles in a set.
  * Then you can iterate over the set and create a dictionary for each allele with `{name: allele_name}`
* [x] Continue on the function that produces `alleles.json`:
  * Iterate along the set to create the complete dictionary (call `replace_allele_features` on each allele, and generate a dictionary like the one in the example)
  * Don't forget to also extract the mating type.

**Comment by manulera** [comment]
Some comments on the current implementation:

* [ ] For json, use lowercase keys, and use underscore instead of spaces: `Mating type -> mating_type`. It's more pythonic to do it this way.
* [ ] You should not use the genotype as the key to the dictionaries, since two strains may have the same genotype. Use the unique identifier of strains instead (`id` / `Sample Name`).
https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/second_version_pipeline.py#L13-L14
* [ ] Here you are not handling  the special cases
https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/second_version_pipeline.py#L9
Remember this:
https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/first_version_pipeline.py#L26
Instead of having to remember this every time, you can instead create a function in the module to load the tsv files (call it `read_strains_tsv` or something like this), we might have to add more arguments or post-processing of the text in the future, so it's good to have it already in a function

More comments incoming, but adding this already

**Comment by manulera** [comment]
For the function `strains_lists`:

A function name should always contain a verb, `strains_list` sounds like a variable name.

You first iterate over genotypes here to extract the mating type and store it in a dictionary, and you also extract the alleles as a list into a dictionary.
https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/second_version_pipeline.py#L14-L25

This line is a bit convoluted:

https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/second_version_pipeline.py#L27

Then you iterate here again:

https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/second_version_pipeline.py#L31

A simpler way of doing this would be (you have to add some stuff still, but you can see it's a bit easier to follow):

```python
# The new function you have to add
from genestorian_module import read_strains_tsv

import re
# A function name should always contain a verb.
# strains_list sounds like a variable name

def build_strain_list(strain_csv_file):
    strain_list = list()
    data = read_strains_tsv(strain_csv_file)

    # Iterate over rows
    for row_index, strain in data.iterrows():
        alleles = list()
        mating_type = 'h?'  # use this as empty value
        for allele in re.split("\s+", strain['Genotype']):
            if allele in ['h90', 'h-', 'h+']:
                mating_type = allele
            else:
                alleles.append(allele)

        strain_list.append({
            'id': strain['Sample Name'],
            # and the other fields
        })
    return strain_list
```

**Comment by manulera** [comment]
> Instead of having to remember this every time, you can instead create a function in the module to load the tsv files (call it read_strains_tsv or something like this), we might have to add more arguments or post-processing of the text in the future, so it's good to have it already in a function

* [ ] This function should also call the method `.astype(str)` on the columns, so that we don't have to do that either.

**Comment by manulera** [comment]
For the function `strains_lists`:

A function name should always contain a verb, `strains_list` sounds like a variable name.

You first iterate over genotypes here to extract the mating type and store it in a dictionary, and you also extract the alleles as a list into a dictionary.
https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/second_version_pipeline.py#L14-L25

This line is a bit convoluted:

https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/second_version_pipeline.py#L27

Then you iterate here again:

https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/second_version_pipeline.py#L31

A simpler way of doing this would be (you have to add some stuff still, but you can see it's a bit easier to follow):

```python
# The new function you have to add
from genestorian_module import read_strains_tsv

import re
# A function name should always contain a verb.
# strains_list sounds like a variable name

def build_strain_list(strain_csv_file):
    strain_list = list()
    data = read_strains_tsv(strain_csv_file)

    # Iterate over rows
    for row_index, strain in data.iterrows():
        alleles = list()
        mating_type = 'h?'  # use this as empty value
        for allele in re.split("\s+", strain['Genotype']):
            if allele in ['h90', 'h-', 'h+']:
                mating_type = allele
            else:
                alleles.append(allele)

        strain_list.append({
            'id': strain['Sample Name'],
            # and the other fields
        })
    return strain_list
```

**Comment by manulera** [comment]
https://github.com/manulera/genestorian_data_refinement/blob/704b546494d662bb79c6600eebc5cff8bc2be842/Lab_strains/dey_lab/second_version_pipeline.py#L97-L102

No need to read the `strains.tsv` file again, you already have splitted the genotypes into alleles, and you have already excluded the mating types, what you can do  instead:

```python
# Create a list of lists, in which each element is the list contained in 'alleles' for each strain.
list_of_lists_allele_names = [genotype['alleles'] for genotype in genotype_list]
# Combine the list into a single list with sum, and then convert it into a set to eliminate duplicates
set(sum(list_of_lists_allele_names,[]))
```

**Comment by manulera** [comment]
* [ ] The `replace_word` parameter that you pass to functions `replaced_allele_feature_name` and `replaced_allele_feature` is not necessary, it can be extracted from the dictionary, and made uppercase. You can make the function `feature_dict` return `feature_type_name` as well, and then make it uppercase.

**Comment by anamika-yadav99** [comment]
- [ ] 'Δ' is read as 'u03b4' when reading from strains.tsv.

**Comment by manulera** [comment]
Great job! Closed with #28 🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀
